### ISSUE TITLE
No way to Close/Dispose WCF Client in a Sync scenario

### ISSUE THREAD

**Issue by lemoinem** [dotnet/wcf#3676]
Following https://github.com/dotnet/wcf/issues/654, We now have synchronous way to call a WCF Service.

However, the Client class still only has `OpenAsync` and `CloseAsync`. In particular, it is still not `Disposable`.

In our case, we have heavyweight use of WCF clients. And this leads to resource exhaustion if we cannot `await CloseAsync()`. Since we have a lot of code, it would be much better to have a fully-Sync and Disposable-safe SoapClient implementation.
Wouldn't a better option be to have two separate classes (one Async, that wouldn't be Disposable, and one Sync, that would be Disposable)?

It seems easier to migrate previous code this way.

Note: We could do `CloseAsync().Wait()`, but that's not a satisfactory solution (https://msdn.microsoft.com/en-us/magazine/jj991977.aspx). In our case, we a migrate part of an App to ASP.Net Core. Our Class Libraries will be built and run both in the context of .Net 4.6.1/ASP.Net and .Net Core 2.2/Asp.Net Core. In addition to some GUI Application. Using `.Wait()` is begging for deadlock.

**Comment by mconnew** [comment]
I would suggest that we should look into the option of implementing System.IAsyncDisposable.

**Comment by lemoinem** [comment]
Would this work with .Net Framework applications?
`IAsyncDisposable` is planed to be part of .Net Standard 2.1 (https://docs.microsoft.com/en-us/dotnet/api/system.iasyncdisposable?view=netstandard-2.1&viewFallbackFrom=netframework-4.8).
.Net Core 3 seems to be planned as supporting it as well (https://docs.microsoft.com/en-us/dotnet/api/system.iasyncdisposable?view=netstandard-2.1&viewFallbackFrom=netframework-4.8) but I couldn't find which version of .Net Framework will support `IAsyncDisposable` or .Net Standard 2.1.

Moreover, as far as I can tell (https://github.com/dotnet/roslyn/issues/114), it wouldn't allow us to keep our code Sync (which is **100%** of our use case). `IAsyncDisposable` implies having async surrounding code (because of the elided `await DisposeAsync();`).

Only having an Sync implementation of `IDisposable` available would fit our use case.
If mixing Async calls+`IDisposable` or Sync calls+`IAsyncDisposable` becomes an issue, generating two Client classes instead of a single one could be an easy solution.

Outside of the Sync use-case, I do agree `IAsyncDisposable` would be a great addition.

**Comment by mconnew** [comment]
You mentioned adding the IDisposable interface, so I was just throwing in there that it should be an option (and only an option because of the netstandard 2.1 issue) to add that too. We could always generate both methods, Dispose and DisposeAsync but only add the interface as an option.  
As to your problem, you don't actually need a Dispose method created for you. If you look at the implementation of the generated CloseAsync method, it looks like this:
```c#
  return System.Threading.Tasks.Task.Factory.FromAsync(((System.ServiceModel.ICommunicationObject)(this)).BeginClose(null, null), new System.Action<System.IAsyncResult>(((System.ServiceModel.ICommunicationObject)(this)).EndClose));
```

The generated client derives from `ClientBase<TChannel>` which implements `ICommunicationObject`. The implementation of CloseAsync uses a Task helper class to connect the two APM methods together to produce a Task. This is needed because `ICommunicationObject` doesn't have a CloseAsync method. `ClientBase<TChannel>` is sort of implementing `ICommunicationObject` explicitly (it's not really, it's an artifact of reference assemblies but just pretend it is and it will have predictable behavior) so you can't directly call the methods, you need to cast it to `ICommunicationObject` first. So in your code, all you need to do is this:
```c#
    ((ICommunicationObject)clientChannel).Close();
```

**Comment by lemoinem** [comment]
Thanks.
I was checking at this `ICommunicationObject.Close()` but was thoroughly confused by the fact that the method wasn't found on .Net Standard and that the interface didn't seem to exist in .Net Standard or Core. I will use this.

Having a Dispose method would still help by allowing us to use `using () { }` blocks. As we were doing with the code generated by VS2015 and previous. Right now, we will have to change all our many calls to use the `try { } finally { }` pattern... (Or most probably, inject the interface and method in our generate `Reference.cs`, as this will allow us to reduce code modifications).

**Comment by mconnew** [comment]
You shouldn't use WCF with a using block anyway as WCF doesn't strictly keep to the Dispose pattern. In WCF, Dispose is basically an alias for Close and that can throw an exception. If it does throw, then you should call abort. Additionally, you shouldn't call Close/Dispose on a CommunicationObject which isn't in the Opened state, instead you should call Abort. Here's an [example](https://github.com/dotnet/wcf/blob/master/src/System.Private.ServiceModel/tests/Common/Scenarios/ScenarioTestHelpers.cs#L184) of how to properly close a WCF communication object.

**Comment by lemoinem** [comment]
Thanks for the example. It very helpful. I completely agree with your point: "It shouldn't be used that way". I'm also aware of the the finer (and probably more to the point) discussion at #26. I'm absolutely not arguing against these.

However, my current issue is not "fixing the way we use WCF". It's migrating **some** parts of a 12 years-old, currently in production worldwide, MLOC software from .Net Framework to .Net Standard/Core.

The current code uses the `using() { }` (anti-)pattern. If the code generated by the new svcutil tool were to implement `IDisposable`, this would mean minor to no modifications required to migrated the code (Client/Service Class renames, at most).
This is a very nice migration path. It would allow us to keep the .Net Framework code to using the old generated code. While the .Net Core code would use the newly generated code. The code in question has been running just fine for over 10 years. We would very much like it to stay that way.

As long as the tool does not generate code implementing `IDisposable`, this means much bigger code changes (and therefore introducing much bigger risks and incurring huge testing costs).

I do not expect the migration to fix our anti-pattern, but at least, not to introduce bigger risks than strictly necessary. (Although having a proper implementation of `IDisposable`, if possible at all, would be great, but again, I believe that discussion is better had in #26).

**Comment by mconnew** [comment]
Okay, I understand your issue now. It's my mistake, I should have asked what your failing scenario was and was focusing on the wrong thing. Are you aware that the code which is generated is a partial class? In your own code you can add the IDisposable interface yourself. Simple add a .cs file to your project containing something similar to this (with the correct class name):
```c#
public partial class MyServiceClient : IDisposable
{
    public void Dispose()
    {
        ((ICommunicationObject)this).Close();
    }
}
```

Now the class implements IDisposable and your existing code should work.

**Comment by lemoinem** [comment]
No problem. Thank you for your patience.
Retrospectively, I realize, I didn't phrase my issue as best as could be.

I believe using the partial class is the way to go here...
I had noticed the partial class, but got confused mid-way by the `ICommunicationObject`'s Schrodinger implementation and then I tried to apply the partial class to the wrong class (I tried to apply it to ClientBase directly...).

The solution may not be as seamless as I would have like. But that'll be good enough (in particular for my stated goal :) ).
I'll probably use an implementation of Dispose closer to the example you linked earlier...
Thanks again.
